apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/Tracer.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.tracer;
+
+import co.elastic.apm.agent.tracer.dispatch.BinaryHeaderGetter;
+import co.elastic.apm.agent.tracer.dispatch.HeaderGetter;
+import co.elastic.apm.agent.tracer.dispatch.TextHeaderGetter;
+import co.elastic.apm.agent.tracer.pooling.ObjectPoolFactory;
+
+import javax.annotation.Nullable;
+
+public interface Tracer {
+
+    boolean isRunning();
+
+    @Nullable
+    <T extends Tracer> T probe(Class<T> type);
+
+    <T extends Tracer> T require(Class<T> type);
+
+    <T> T getConfig(Class<T> configuration);
```

Comment:
The method is not really nullable, though. For example:

tracer.getConfig(JaxRsConfiguration.class).isUseJaxRsPathForTransactionName();

I would rather have implementations throw an explicit exception if the requested configuration class is not available.

Reply:
Good point, this make it simpler on the caller side to not have to deal with potential `null` values.

However in this case we'd better make `co.elastic.apm.agent.impl.ElasticApmTracer#getConfig` throw an exception rather than return `null`, also the stagemonitor implementation can also return `null` in `co.elastic.apm.agent.impl.ElasticApmTracer#getConfig`. This is definitely not related to your changes here, but while we are changing this, why not take the opportunity to clean-up. If you don't mind I'll push a commit directly to this PR to change this.

Reply:
I added the change.

Another upside of this would be that one would get an exception if requesting a configuration before the tracer is initialized. If `null` was a legitimate return value, the noop tracer would currently have triggered this default configuration, even if an actual configuration becomes available later.